infer/component.go

```diff
@@ -98,3 +100,16 @@ func (rc *derivedComponentController[R, I, O]) Construct(pctx p.Context, typ str
 	}
 	return res, err
 }
+
+type componentContextKey struct{}
+
+// Retrieve a provider.Context from a pulumi.Context.
+//
+// This function is only valid when the *pulumi.Context was passed in from the infer
+// library.
+func CtxFromPulumiContext(ctx *pulumi.Context) p.Context {
```

Comment:
Should we move this to be next to the other `CtxXXX` funcs in `provider.go`?

That should avoid the need to have `embeddedData`: we wouldn't have to publicly export `GetEmbeddedData` and `PutEmbeddedData`, and wouldn't have to have `putEmbeddedMap` sprinkled throughout.

See my comment below on the implementation of `Construct` in `provider.go` below.